### PR TITLE
[search] Get rid of duplication in the downloader search.

### PR DESCRIPTION
**Pull request by mpimenov**
@ygorshenin PTAL

### CODE REVIEW

**Comment by mpimenov** [comment]
DUDE JTTP

File: search/downloader_search_callback.cpp [hunk 1]
```diff
@@ -0,0 +1,108 @@
...
+                                  string & name)
+{
+  int8_t langIndices[] = {StringUtf8Multilang::kEnglishCode, StringUtf8Multilang::kDefaultCode,
+                          StringUtf8Multilang::kInternationalCode};
+
+  for (auto const langIndex : langIndices)
+  {
+    if (!ft.GetName(langIndex, name))
+      continue;
+    if (storage.IsCoutryIdCountryTreeInnerNode(name))
```
Comment (ygorshenin) [inline search/downloader_search_callback.cpp]: s/Coutry/Country/g

File: search/search_integration_tests/downloader_search_test.cpp [hunk 2]
```diff
@@ -0,0 +1,231 @@
...
+  {
+  }
+};
+}  // namespace storage
+
+namespace search
+{
+namespace
+{
+string const kCountriesTxt = string(R"({
```
Comment (ygorshenin) [inline search/search_integration_tests/downloader_search_test.cpp]: No need in string surrounding R"..."?

File: search/search_integration_tests/downloader_search_test.cpp [hunk 3]
```diff
@@ -0,0 +1,231 @@
...
+
+#include "std/algorithm.hpp"
+#include "std/string.hpp"
+
+using namespace generator::tests_support;
+using namespace search::tests_support;
+
+using TRules = vector<shared_ptr<MatchingRule>>;
+
+namespace storage
```
Comment (ygorshenin) [inline search/search_integration_tests/downloader_search_test.cpp]: Why? Just put this class to the anonymous namespace. No need to extend storage namespace.

File: search/search_integration_tests/downloader_search_test.cpp [hunk 4]
```diff
@@ -0,0 +1,231 @@
...
+#include "base/logging.hpp"
+#include "base/macros.hpp"
+
+#include "std/algorithm.hpp"
+#include "std/string.hpp"
+
+using namespace generator::tests_support;
+using namespace search::tests_support;
+
+using TRules = vector<shared_ptr<MatchingRule>>;
```
Comment (ygorshenin) [inline search/search_integration_tests/downloader_search_test.cpp]: I guess it's time to move this using to test_results_matching.

File: search/search_integration_tests/downloader_search_test.cpp [hunk 5]
```diff
@@ -0,0 +1,231 @@
...
+  }
+
+  void OnResultsDownloader(search::Results const & results)
+  {
+    OnResults(results);
+    m_downloaderCallback(results);
+  }
+
+  vector<storage::DownloaderSearchResult> const & GetResults() const { return m_downloaderResults; }
+private:
```
Comment (ygorshenin) [inline search/search_integration_tests/downloader_search_test.cpp]: Please, insert a blank line.

File: search/search_integration_tests/downloader_search_test.cpp [hunk 6]
```diff
@@ -0,0 +1,231 @@
...
+    p.m_onResults = [this](storage::DownloaderSearchResults const & r)
+    {
+      CHECK(!m_endMarker, ());
+      if (r.m_endMarker)
+      {
+        m_endMarker = true;
+        return;
+      }
+
+      m_downloaderResults.insert(m_downloaderResults.end(), r.m_results.begin(), r.m_results.end());
```
Comment (ygorshenin) [inline search/search_integration_tests/downloader_search_test.cpp]: Data race on m_downloaderResults here.

File: search/search_integration_tests/downloader_search_test.cpp [hunk 7]
```diff
@@ -0,0 +1,231 @@
...
+    p.SetForceSearch(true);
+    return p;
+  }
+
+  storage::DownloaderSearchParams MakeDownloaderParams(string const & query)
+  {
+    storage::DownloaderSearchParams p;
+    p.m_query = query;
+    p.m_inputLocale = "en";
+    p.m_onResults = [this](storage::DownloaderSearchResults const & r)
```
Comment (ygorshenin) [inline search/search_integration_tests/downloader_search_test.cpp]: I guess you need to call TestSearchRequest::OnResults here, instead of this.

Comment (ygorshenin) [inline search/search_integration_tests/downloader_search_test.cpp]: It should accumulate on results, and, moreover, there won't be any data race on m_downloaderResults.

File: search/downloader_search_callback.cpp (hunk 1 above)
Comment (mpimenov) [inline search/downloader_search_callback.cpp]: done

File: search/search_integration_tests/downloader_search_test.cpp (hunk 2 above)
Comment (mpimenov) [inline search/search_integration_tests/downloader_search_test.cpp]: done

File: search/search_integration_tests/downloader_search_test.cpp (hunk 3 above)
Comment (mpimenov) [inline search/search_integration_tests/downloader_search_test.cpp]: done

File: search/search_integration_tests/downloader_search_test.cpp (hunk 4 above)
Comment (mpimenov) [inline search/search_integration_tests/downloader_search_test.cpp]: In fact, I don't need TRules here at all, but why not?
Moved.

File: search/search_integration_tests/downloader_search_test.cpp (hunk 5 above)
Comment (mpimenov) [inline search/search_integration_tests/downloader_search_test.cpp]: done

File: search/search_integration_tests/downloader_search_test.cpp (hunk 6 above)
Comment (mpimenov) [inline search/search_integration_tests/downloader_search_test.cpp]: Nice catch.
done

File: search/search_integration_tests/downloader_search_test.cpp (hunk 7 above)
Comment (mpimenov) [inline search/search_integration_tests/downloader_search_test.cpp]: Note that ```DownloaderSearchResults``` are different from the search ```Results```.

**Comment by mpimenov** [comment]
PTAL

File: map/framework.hpp [hunk 8]
```diff
@@ -71,6 +72,7 @@ struct ViewportSearchParams;
 namespace storage
 {
 class CountryInfoGetter;
+struct DownloaderSearchParams;
```
Comment (ygorshenin) [inline map/framework.hpp]: Why do you need it there?

File: search/downloader_search_callback.cpp [hunk 9]
```diff
@@ -0,0 +1,108 @@
...
+
+#include "base/logging.hpp"
+#include "base/string_utils.hpp"
+
+namespace
+{
+bool GetGroupCountryIdFromFeature(storage::Storage const & storage, FeatureType const & ft,
+                                  string & name)
+{
+  int8_t langIndices[] = {StringUtf8Multilang::kEnglishCode, StringUtf8Multilang::kDefaultCode,
```
Comment (ygorshenin) [inline search/downloader_search_callback.cpp]: const

File: search/downloader_search_callback.hpp [hunk 10]
```diff
@@ -0,0 +1,52 @@
...
+public:
+  class Delegate
+  {
+  public:
+    virtual ~Delegate() = default;
+
+    virtual void RunUITask(function<void()> fn) = 0;
+  };
+
+  using TOnResults = storage::DownloaderSearchParams::TOnResults;
```
Comment (ygorshenin) [inline search/downloader_search_callback.hpp]: Do you need this?

File: search/search_integration_tests/downloader_search_test.cpp [hunk 11]
```diff
@@ -0,0 +1,227 @@
...
+      {
+        // Add the country feature to one region only.
+        builder.Add(country);
+        worldCountries.push_back(country);
+      }
+    });
+    worldCities.push_back(capital);
+  };
+
+  addRegion("Flatland", "Square One", m2::PointD(0.0, 0.0), m2::PointD(1.0, 1.0));
```
Comment (ygorshenin) [inline search/search_integration_tests/downloader_search_test.cpp]: Could you please add a test that searches "Square"?

File: search/downloader_search_callback.cpp (hunk 9 above)
Comment (mpimenov) [inline search/downloader_search_callback.cpp]: done

File: search/downloader_search_callback.hpp (hunk 10 above)
Comment (mpimenov) [inline search/downloader_search_callback.hpp]: Not really.
done

File: map/framework.hpp (hunk 8 above)
Comment (mpimenov) [inline map/framework.hpp]: Now that I've added the delegate, I don't.
done

Comment (mpimenov) [inline map/framework.hpp]: On second thought, I've left it as is and removed the explicit include.

File: search/search_integration_tests/downloader_search_test.cpp (hunk 11 above)
Comment (mpimenov) [inline search/search_integration_tests/downloader_search_test.cpp]: That's for another PR.
I can add it with a todo.

**Comment by ygorshenin** [comment]
LGTM

**Pull request merged**